pkg/dmaasbackup/nonperiodic_backup.go

```diff
@@ -121,44 +122,89 @@ func (d *dmaasBackup) cleanupNonPeriodicSchedule(dbkp *v1alpha1.DMaaSBackup) err
 		return nil
 	}
 
-	backupList, err := d.backupLister.List(
-		labels.SelectorFromSet(map[string]string{
-			v1alpha1.DMaaSBackupLabelKey:  dbkp.Name,
-			velerov1api.ScheduleNameLabel: schedule.ScheduleName,
-		}),
-	)
-
+	backupList, err := d.listScheduledBackups(dbkp, schedule.ScheduleName)
 	if err != nil {
 		return errors.Wrapf(err, "failed to get list of backup for schedule=%s", schedule.ScheduleName)
 	}
 
+	if len(backupList) < retainCount {
+		d.logger.Debugf("Number of backups are %v, required %v backups to trigger cleanup",
+			len(backupList),
+			retainCount)
+		return nil
+	}
+
 	defer d.logger.Debug("Completed cleanup for non-periodic schedule")
 
 	sort.Sort(sort.Reverse(byBackupCreationTimeStamp(backupList)))
 
-	// We need to retain `fullBackupRetentionThreshold` number of backups
-	// check for completed backup count
-	var completedBackup int
+	requiredCompletedBackup := retainCount
+	completedIdx, successfulIdx := getBackupCleanupIdx(backupList, &requiredCompletedBackup)
+	if completedIdx == -1 {
+		d.logger.Debugf("Number of completed Backups is %v, required %v completed backups to trigger cleanup",
+			requiredCompletedBackup,
+			retainCount)
+		return nil
+	}
+
+	if !dbkp.Spec.PeriodicFullBackupCfg.DisableSuccessfulBackupRetain {
+		if successfulIdx == -1 {
+			d.logger.Debugf("No successful backup exists, cleanup skipped")
+			return nil
+		}
+
+		if completedIdx <= successfulIdx {
```

Comment:
some comments here please for the logic

Reply:
adding a print here would be useful. This will avoid flooding by adding logs at other places.

Reply:
also, a log when things related to retention are fine, i.e., from hitting this case to the case where retention count is working fine.
its kind of logging when the edge is triggered.

Reply:
ok